homeassistant/components/adaptive_lighting/switch.py

```diff
@@ -0,0 +1,792 @@
+
+    async def _setup_listeners(self, _=None):
+        _LOGGER.debug("%s: Called '_setup_listeners'", self._name)
+        if not self.is_on or not self.hass.is_running:
+            _LOGGER.debug("%s: Cancelled '_setup_listeners'", self._name)
+            return
+        assert not self.remove_listeners
+        remove_interval = async_track_time_interval(
+            self.hass, self._async_update_at_interval, self._interval
+        )
+        self.remove_listeners.append(remove_interval)
+        if self._lights:
+            self._expand_light_groups()
+            remove_state = async_track_state_change_event(
+                self.hass, self._lights, self._light_event
+            )
+            self.remove_listeners.append(remove_state)
+        if self._sleep_entity is not None:
+            remove_sleep = async_track_state_change_event(
+                self.hass,
+                self._sleep_entity,
+                self._sleep_state_event,
+            )
+            self.remove_listeners.append(remove_sleep)
+        if self._disable_entity is not None:
+            remove_disable = async_track_state_change_event(
+                self.hass,
+                self._disable_entity,
+                self._disable_state_event,
+            )
+            self.remove_listeners.append(remove_disable)
+
+    def _remove_listeners(self):
+        while self.remove_listeners:
+            remove_listener = self.remove_listeners.pop()
+            remove_listener()
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return self._icon
+
+    @property
+    def device_state_attributes(self):
+        """Return the attributes of the switch."""
+        attrs = {
+            "percent": self._percent,
+            "brightness": self._brightness,
+            "color_temp_kelvin": self._color_temp_kelvin,
+            "color_temp_mired": self._color_temp_mired,
+            "rgb_color": self._rgb_color,
+            "xy_color": self._xy_color,
+            "hs_color": self._hs_color,
+        }
+        if not self.is_on:
+            return {key: None for key in attrs}
+        return attrs
+
+    async def async_turn_on(
+        self, adapt_lights: bool = True
+    ):  # pylint: disable=arguments-differ
+        """Turn on adaptive lighting."""
+        _LOGGER.debug(
+            "%s: Called 'async_turn_on', current state is '%s'", self._name, self._state
+        )
+        if self.is_on:
+            return
+        self._state = True
+        await self._setup_listeners()
+        if adapt_lights:
+            await self._maybe_adapt_lights(
+                transition=self._initial_transition, force=True
+            )
+
+    async def async_turn_off(self, **kwargs):
+        """Turn off adaptive lighting."""
+        if not self.is_on:
+            return
+        self._state = False
+        self._remove_listeners()
+
+    async def _update_attrs(self):
```

Comment:
It looks like this could be a callback as nothing is being awaited here

Reply:
Thanks @bdraco, fixed in 6027404c91.